File: feishu_builder_agent/llm_client.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Protocol
from urllib import error, request

from .logging_utils import builder_log


class JsonLLMClient(Protocol):
    def generate_json(self, *, system_prompt: str, user_prompt: str) -> dict[str, Any]:
        """Generate structured JSON output."""


@dataclass(frozen=True)
class DisabledLLMClient:
    """Placeholder client for environments where no live model is configured."""

    reason: str = "LLM client not configured"

    def generate_json(self, *, system_prompt: str, user_prompt: str) -> dict[str, Any]:
        raise RuntimeError(self.reason)


def _read_env_file(path: str | Path) -> dict[str, str]:
    values: dict[str, str] = {}
    env_path = Path(path)
    if not env_path.exists():
        return values
    for line in env_path.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#") or "=" not in stripped:
            continue
        key, raw_value = stripped.split("=", 1)
        values[key.strip()] = raw_value.strip()
    return values


def _load_builder_env() -> dict[str, str]:
    values = dict(os.environ)
    root_env = _read_env_file(".env")
    for key, value in root_env.items():
        values.setdefault(key, value)
    return values


def _normalize_chat_completions_url(base_url: str) -> str:
    base = base_url.rstrip("/")
    if base.endswith("/chat/completions"):
        return base
    return f"{base}/chat/completions"


def _parse_json_object(text: str) -> dict[str, Any]:
    stripped = text.strip()
    if not stripped:
        raise RuntimeError("model returned empty content")
    try:
        value = json.loads(stripped)
    except json.JSONDecodeError:
        start = stripped.find("{")
        end = stripped.rfind("}")
        if start < 0 or end <= start:
            raise RuntimeError("model did not return valid JSON")
        value = json.loads(stripped[start : end + 1])
    if not isinstance(value, dict):
        raise RuntimeError("model JSON response must be an object")
    return value


@dataclass(frozen=True)
class OpenAICompatibleLLMClient:
    api_key: str
    base_url: str
    model: str
    timeout_seconds: float = 120.0
    max_tokens: int = 1200

    def _request(self, payload: dict[str, Any]) -> dict[str, Any]:
        builder_log(
            "llm",
            f"开始请求模型 model={self.model} timeout={self.timeout_seconds}s max_tokens={self.max_tokens}",
        )
        req = request.Request(
            _normalize_chat_completions_url(self.base_url),
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.api_key}",
            },
            method="POST",
        )
        try:
            with request.urlopen(req, timeout=self.timeout_seconds) as response:
                raw = response.read().decode("utf-8", errors="replace")
        except error.HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace")
            builder_log("llm", f"模型请求返回 HTTP {exc.code}")
            raise RuntimeError(f"LLM HTTP {exc.code}: {raw}") from exc
        except error.URLError as exc:
            builder_log("llm", f"模型请求网络失败: {exc}")
            raise RuntimeError(f"LLM request failed: {exc}") from exc
        payload_json = json.loads(raw)
        choices = payload_json.get("choices")
        if not isinstance(choices, list) or not choices:
            raise RuntimeError("LLM response missing choices")
        message = choices[0].get("message") if isinstance(choices[0], dict) else None
        content = message.get("content") if isinstance(message, dict) else None
        if isinstance(content, list):
            parts = []
            for item in content:
                if isinstance(item, dict) and item.get("type") == "text":
                    parts.append(str(item.get("text") or ""))
            content = "".join(parts)
        builder_log("llm", "模型请求成功返回 JSON 内容")
        return _parse_json_object(str(content or ""))

    def generate_json(self, *, system_prompt: str, user_prompt: str) -> dict[str, Any]:
        payload = {
            "model": self.model,
            "temperature": 0.4,
            "max_tokens": self.max_tokens,
            "response_format": {"type": "json_object"},
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
        }
        try:
            return self._request(payload)
        except RuntimeError as exc:
            text = str(exc)
            if "response_format.type" not in text and "json_object" not in text:
                raise
            builder_log("llm", "当前模型不支持 json_object，准备回退到普通 JSON 提示词模式")
        fallback_payload = {
            "model": self.model,
            "temperature": 0.2,
            "max_tokens": self.max_tokens,
            "messages": [
                {
                    "role": "system",
                    "content": system_prompt
                    + "\nReturn one JSON object only. Do not wrap it in Markdown. Do not add commentary.",
                },
                {"role": "user", "content": user_prompt},
            ],
        }
        return self._request(fallback_payload)


def build_llm_client_from_env() -> JsonLLMClient:
    env = _load_builder_env()
    api_key = str(env.get("OPENAI_API_KEY") or "").strip()
    base_url = str(env.get("OPENAI_API_BASE_URL") or "").strip()
    model = (
        str(env.get("FEISHU_BUILDER_MODEL") or "").strip()
        or str(env.get("VOLCENGINE_EP_ID") or "").strip()
        or str(env.get("OPENAI_MODEL") or "").strip()
    )
    timeout_seconds = float(str(env.get("FEISHU_BUILDER_TIMEOUT_SECONDS") or "120").strip() or "120")
    max_tokens = int(str(env.get("FEISHU_BUILDER_MAX_TOKENS") or "1200").strip() or "1200")
    if not api_key or not base_url or not model:
        builder_log("llm", "未检测到可用的 OPENAI 兼容配置，当前将使用 fallback 生成链")
        return DisabledLLMClient("OPENAI-compatible builder model is not configured")
    builder_log("llm", f"已加载 OPENAI 兼容配置 model={model} base_url={base_url}")
    return OpenAICompatibleLLMClient(
        api_key=api_key,
        base_url=base_url,
        model=model,
        timeout_seconds=timeout_seconds,
        max_tokens=max_tokens,
    )

File: extensions/feishu-task-wiki/openclaw-lark/src/task-events/llm-client.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");

function readEnvFile(filePath) {
  try {
    const raw = fs.readFileSync(filePath, "utf8");
    const values = {};
    for (const line of raw.split(/\r?\n/u)) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#") || !trimmed.includes("=")) {
        continue;
      }
      const separator = trimmed.indexOf("=");
      const key = trimmed.slice(0, separator).trim();
      const value = trimmed.slice(separator + 1).trim();
      if (!key) {
        continue;
      }
      values[key] = value.replace(/^['"]|['"]$/g, "");
    }
    return values;
  } catch {
    return {};
  }
}

function loadEnv() {
  const merged = { ...process.env };
  const rootEnv = readEnvFile(path.join(process.cwd(), ".env"));
  for (const [key, value] of Object.entries(rootEnv)) {
    if (!merged[key]) {
      merged[key] = value;
    }
  }
  return merged;
}

function normalizeChatCompletionsUrl(baseUrl) {
  const base = String(baseUrl ?? "").trim().replace(/\/+$/u, "");
  if (!base) {
    return "";
  }
  if (base.endsWith("/chat/completions")) {
    return base;
  }
  return `${base}/chat/completions`;
}

function parseJsonObject(text) {
  const raw = String(text ?? "").trim();
  if (!raw) {
    throw new Error("model returned empty content");
  }
  try {
    const parsed = JSON.parse(raw);
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new Error("model JSON response must be an object");
    }
    return parsed;
  } catch {
    const start = raw.indexOf("{");
    const end = raw.lastIndexOf("}");
    if (start < 0 || end <= start) {
      throw new Error("model did not return valid JSON");
    }
    const parsed = JSON.parse(raw.slice(start, end + 1));
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new Error("model JSON response must be an object");
    }
    return parsed;
  }
}

function buildClientFromEnv() {
  const env = loadEnv();
  const apiKey = String(env.FEISHU_TASK_WIKI_API_KEY ?? env.OPENAI_API_KEY ?? "").trim();
  const baseUrl = String(env.FEISHU_TASK_WIKI_API_BASE_URL ?? env.OPENAI_API_BASE_URL ?? "").trim();
  const model = String(
    env.FEISHU_TASK_WIKI_MODEL
      ?? env.FEISHU_BUILDER_MODEL
      ?? env.VOLCENGINE_EP_ID
      ?? env.OPENAI_MODEL
      ?? "",
  ).trim();
  const timeoutMsValue = env.FEISHU_TASK_WIKI_TIMEOUT_MS;
  const timeoutSecondsValue = env.FEISHU_BUILDER_TIMEOUT_SECONDS;
  const timeoutMs = timeoutMsValue
    ? Number.parseInt(String(timeoutMsValue), 10)
    : timeoutSecondsValue
      ? Number.parseInt(String(timeoutSecondsValue), 10) * 1000
      : 120000;
  const maxTokens = Number.parseInt(
    String(env.FEISHU_TASK_WIKI_MAX_TOKENS ?? env.FEISHU_BUILDER_MAX_TOKENS ?? "1600"),
    10,
  );
  if (!apiKey || !baseUrl || !model) {
    return null;
  }
  return {
    apiKey,
    baseUrl: normalizeChatCompletionsUrl(baseUrl),
    model,
    timeoutMs: Number.isFinite(timeoutMs) ? timeoutMs : 120000,
    maxTokens: Number.isFinite(maxTokens) ? maxTokens : 1600,
  };
}

async function requestJson(params) {
  const client = buildClientFromEnv();
  if (!client) {
    throw new Error("task wiki LLM client is not configured");
  }
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), client.timeoutMs);
  const payload = {
    model: client.model,
    temperature: params.temperature ?? 0.2,
    max_tokens: params.maxTokens ?? client.maxTokens,
    response_format: { type: "json_object" },
    messages: [
      { role: "system", content: params.systemPrompt },
      { role: "user", content: params.userPrompt },
    ],
  };
  try {
    const response = await fetch(client.baseUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${client.apiKey}`,
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });
    const raw = await response.text();
    if (!response.ok) {
      throw new Error(`LLM HTTP ${response.status}: ${raw}`);
    }
    const parsed = JSON.parse(raw);
    const choice = Array.isArray(parsed?.choices) ? parsed.choices[0] : null;
    const message = choice?.message;
    let content = message?.content ?? "";
    if (Array.isArray(content)) {
      content = content
        .filter((item) => item && item.type === "text")
        .map((item) => String(item.text ?? ""))
        .join("");
    }
    return parseJsonObject(content);
  } catch (error) {
    const message = String(error?.message ?? error);
    if (!/json_object|response_format/u.test(message)) {
      throw error;
    }
    const fallbackPayload = {
      model: client.model,
      temperature: params.temperature ?? 0.2,
      max_tokens: params.maxTokens ?? client.maxTokens,
      messages: [
        {
          role: "system",
          content: `${params.systemPrompt}\nReturn one JSON object only. Do not use Markdown.`,
        },
        { role: "user", content: params.userPrompt },
      ],
    };
    const response = await fetch(client.baseUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${client.apiKey}`,
      },
      body: JSON.stringify(fallbackPayload),
      signal: controller.signal,
    });
    const raw = await response.text();
    if (!response.ok) {
      throw new Error(`LLM HTTP ${response.status}: ${raw}`);
    }
    const parsed = JSON.parse(raw);
    const choice = Array.isArray(parsed?.choices) ? parsed.choices[0] : null;
    return parseJsonObject(choice?.message?.content ?? "");
  } finally {
    clearTimeout(timeout);
  }
}

module.exports = {
  buildClientFromEnv,
  requestJson,
};
